cmd/k8s-operator: check that cluster traffic is routed to egress ProxyGroup Pod before marking it as ready

irbekrm opened the pull request:
This change is part of work to minimize downtime when updates are applied to egress `ProxyGroup` that require restart, see https://github.com/tailscale/tailscale/issues/14326.
It adds a `tailscale.com/egress-services` [readiness gate](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-readiness-gate) to `ProxyGroup`  `Pod`s. The corresponding readiness condition is set by a new `egress-pods-readiness-reconciler` reconciler that uses the health check endpoint on the egress services to test whether the `Pod` is set up as a backend for the egress ClusterIP Service.

## Problem

When an update is applied to `ProxyGroup` that require `Pod` restart, this is done using [`StatefulSet` rolling update](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#rolling-updates):

1. `Pod` N gets terminated
2. `Pod` N gets recreated with a new IP address
3. `Pod` N is marked ready
4. When `Pod` N is marked ready, steps 1 - 3 are repeated for Pod N-1
(in a separate flow):
5. [`egress-eps-reconciler`](https://github.com/tailscale/tailscale/blob/main/cmd/k8s-operator/egress-eps.go) is triggered on `Pod` changes and reconciles all `EndpointSlice`s for the `ProxyGroup` removing the endpoints for deleted `Pod` IPs and adding endpoints for the new IPs
6. kube proxy reconciles updated `EndpointSlice`s and updates cluster routing rules, so that cluster traffic for egress endpoints is routed to the new `Pod`s.

Because the steps 1 - 4 for all `Pod`s happen in a separate flow from steps 5 - 6, it is easy to end up in a situtation where all `Pod`s have been recreated with new IPs, but none of the new IPs are yet set up as an endpoint for the egress service (especially if the replica count is just two).

## Solution

So the new sequence when a Pod is recreated now is:
1. A `Pod` starts, it reads the egress services config, sets up netfilter rules to route cluster traffic to tailnet endpoints and updates its status in the `Secret`
2. `egress-eps reconciler` reconciles `EndpointSlice`s on `Pod`/state `Secret` changes, [observes from `Secret` contents that in-proxy routing to tailnet targets has been set up](https://github.com/tailscale/tailscale/blob/bce05ec6c3f3cb2dad1086472e99e2e69b2cfadc/cmd/k8s-operator/egress-eps.go#L153) and updates the `EndpointSlice`s with the Pod's IP
3. kube proxy reconciles the updated `EndpointSlice`s and sets up cluster routes so that traffic for egress services is routed to the new Pod
4. the new `egress-pods-readiness-reconciler` reconciler reconciles the `Pod`, determines that the Pod has been set up as a backend for the egress service and sets the `egress-pods-readiness-reconciler` readiness condition on the `Pod`
5. kubelet sees the `tailscale.com/egress-services` readiness condition and marks the `Pod` as ready
6. once the `Pod` is marked as ready, the next replica can be restarted

## Notes

- It might seem counterintuitive to allow kube proxy to set up routes to Pods that are not 'ready'. However, in practice, reading around about how others solve similar issues, this does not seem to differ from their approach (TODO: links)
The cluster route to an egress target is only set up when the replica's egress service's status confirms that it has set up netfilter routing to the tailnet target, so the Pod can be considered ready once there is a cluster route set up for each egress target.

- The readiness gate logic is not used if `TS_LOCAL_ADDR_PORT` is set - see description of #14326

- This is backwards compatible with older proxy versions

### Alternative implementations

- Instead of operator, the proxy could itself call the health check endpoint (like it already does in the prestop hook) and itself update the Pod status with the readiness condition. I opted for a new operator reconciler because it can be implemented in a cleaner and cheaper way there (no need to add more kube code to shared libraries, update Pod statuses in the shared kube client, etc). The caveat of _this_ implementation is that we now need the operator be up for StatefulSet restarts to succeed. I think this is fine - once we know what the HA for all components look like and how much we want to expand this in the future, we can decide whether we need separate kube proxy images that could pull in kube libraries and perform more kube actions in a safer way.


## Next steps

- This PR + #14326 seem to significantly improve the failover (I've tested this manually on a multi-node cluster), however there is still some downtime caused by delays in netmap update, which needs fixing

tomhjp reviewed (approved):
Nice tests! It took me a while to convince myself that there isn't an alternative solution we should test first, but I think this looks really solid.

tomhjp commented:
> The routes are only set up when the replica has advertised via its status that it has set up netfilter routing to backend, so for routing purposes the Pod can be considered ready at that point.

Where do we manage this? It seems like we'd need a readiness probe for this but we don't define one. I could see that being something we add in future for the next-order-of-magnitude-down errors.

irbekrm commented on cmd/k8s-operator/egress-pod-readiness.go:
```diff
@@ -0,0 +1,274 @@
+	}
+	if rm := routesMissing.Load(); rm {
+		l.Info("Pod is not yet added as an endpoint for all egress targets, waiting...")
+		return reconcile.Result{RequeueAfter: shortRequeue}, nil
```
We could set that condition, but in false state, I think that would maybe still work..
I am not sure about the time- these requeues will carry on 'forever' with a static interval- I am not sure what users could learn from that time? Like, if the condition was false then that would be just what the current state is

irbekrm replied:
Actually, the users will be able to observe the status as kubelet will mark the Pod as 'not ready because readiness gate condition is missing', so we shouldn't need to explicitly do it ourselves

irbekrm commented:
> Where do we manage this? It seems like we'd need a readiness probe for this but we don't define one. I could see that being something we add in future for the next-order-of-magnitude-down errors.

I've updated that comment, it was too high level. I haven't been able to find use for readiness probe for us- we do need to be able to tell when an arbitrary egress service, added to ProxyGroup at any point, has been set up without incorrectly stating that the whole replica is not ready and we need something that works just to gate the initial readiness to control the restart sequence.

irbekrm commented:
> next-order-of-magnitude-down errors

what concrete errors were you thinking?
